documentation/Linting-rules.md

```diff
@@ -0,0 +1,247 @@
+|path-params | Path parameters are correct and valid | Yes| Yes| Error |  |
+|tag-description | global tags have description | No | ???| Warning |  |
+|typed-enum | Enum values should respect the type specifier. | Yes| Yes| Warning |  |
+||||| |
+|oas3-api-servers | OpenAPI servers must be present and non-empty array | Yes| Yes| Warning |  |
+|oas3-examples-value-or-externalValue | Examples for requestBody or response examples can have an `externalValue` or a `value`, but they cannot have both| Yes| Yes| Warning |  |
+|**oas3-operation-security-defined** | Operation `security` values must match a scheme defined in the `components.securitySchemes` object. | Yes| Yes| Warning |  |
+|oas3-parameter-description | Parameter objects should have a description| No| Yes?| Warning |  |
+|oas3-schema | Validate structure of OpenAPI v3 specification | Yes| Yes| Warning |  |
+|oas3-server-not-example.com | Server URL should not point to *example.com*| No| Yes?| Warning |  |
+|oas3-server-trailing-slash | Server URL should not have a trailing slash | Yes| Yes| Warning |  |
+|oas3-unused-component | Potential unused reusable components entry has been detected  | Yes| Yes| Warning |  |
+|oas3-valid-media-example | Examples must be valid against their defined schema. This rule is applied to *Media Type objects*  | Yes| Yes| Warning  |  |
+|oas3-valid-schema-example | Examples must be valid against their defined schema. This rule is applied to *Schema objects* | Yes| Yes| Warning   |  |
+|oas3-server-variables | This rule ensures that server variables defined in OpenAPI Specification 3 (OAS3) and 3.1 are valid, not unused | Yes| Yes| Warning  |  |
+
+
+
+## 2. Language
+### Spell checking
+
+CAMARA API specification files include inline documentation.
+
+The description attributes should be checked for typos.
+
+_Spectral rule_: [camara-language-spelling]()
+
+*Severity*: `warn`
+
+### Reduce telco-specific terminology in API definitions
+
+API Design Guidelines: [2.5 Reduce telco-specific terminology in API definitions](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#25-reduce-telco-specific-terminology-in-api-definitions)
+
+Consider and account for how the API can be fulfilled on a range of network types.
+Avoid terms/types specific to a given telco domain -  terms should be inclusive beyond mobile network. 
+
+See also [CAMARA Glossary](https://github.com/camaraproject/Commonalities/blob/main/documentation/Glossary.md)
+
+
+| ❌ &nbsp; Not recommended | 👍  &nbsp; Recommended |
+|----------------------------|-------------------------|
+| `UE`                 | `device`           |
+| `MSISDN`                 | `phone number`           |
+| `mobile network`      | `network`         |
+
+
+_Spectral rule_: [camara-language-avoid-telco]()
+
+*Severity*: `hint`
+
+## 3. API Definition

```

Comment:
|  &nbsp; Topic|  &nbsp; Description |
|----------------------------|-------------------------|
| Openapi version                | Camara APIs MUST be based on OpenAPI specification with minimum Version 3.0.0.           |
| Request body structure on PUT, POST, PATCH    |    POST / PUT / PATCH methods must have a requestBody. The requestBody content must be "application/json".   |
| Info object: title, version, description   |  "Info" section is mandatory. Title, Description and Version must be documented.Note: It is provided by Spectral.    |
|  Info object: API version format  |   API version within the Info section must follow the format: X.Y.Z   |
| Valid basepath validation   |   API must have valid basepath   |
| Camel case / lower case /Snake_Case   |    Paths URI (tasks, individual resources, etc.) MUST be  lower case and hyphens. OperationIds must be defined in lowerCamelCase Objects must be defined in CamelCase inside properties field. The event type must be written in UPPER_SNAKE_CASE  |
| Path URI   |  URIs must contain the "major version" of the API.    |
| Resource path must not contain the base URL   | Path must not contain the base path.     |
| Sensitive data (msisdn/imsi) cannot be a path or query parameter   |     Sensitive data (msisdn/imsi) cannot be a path or query parameter.Note: Needs to list down if we have other sensitive parameters other than MSISDN/IMSI |
|Valid methods    |     Valid methods method are: GET, PUT, POST, DELETE, PATCH, OPTIONS, HEAD |
|  GET/DELETE methods must not accept a "body"  |  'GET' and 'DELETE' http methods MUST NOT accept a 'requestBody' attribute    |
| Define common error response pattern   |  Following error responses must be define: 400, 401, 403, 404, 405, 406, 429, 500, 501, 502, 503, 504.Note: List needs to be validated as per Camara API documentation    |
|  Verb methods cannot be in the Path  |   Resources must not contain the method name get, put, post, delete, patch.   |
| Parameters must have a description   |   All parameters must have a description.   |
| Path version must be major versions only   |  Path must contain the major API version with format /vX.Example: paths: "/device-status/v0/roaming"    |
|  GET/DELETE methods must not accept a "body"  |   'GET' and 'DELETE' http methods MUST NOT accept a 'requestBody' attribute   |
| Mandatory request body with POST, PATCH and PUT methods   |      The requestBody object MAY be present when the operation is used with one of PUT, POST, PATCH HTTP Verbs.|
|  Summary associated with each operation  |   Summary must be defined on each operation, describing with a short summary what the operation does.   |
| Include valid response examples   |   For each response code, we must define response payload examples   |